### ISSUE TITLE
Cross Repository Push

### ISSUE THREAD

**Issue by jramseye** [docker/distribution#908]
# Specification Proposal: Cross Repository Push
This specification proposal offers a set of changes that can be made to implement cross repository push, suggested in [#634](https://github.com/docker/distribution/issues/634)

In essence, this proposal offers a solution that will improve performance when pushing or pulling from one repository to another.

Consider a situation where User “A” pulls an image and all of its layers from the repository `docker.io/OtherUser/testRepo`. User A then immediately pushes these contents back up to his or her own repository: `docker.io/A/testRepo`

In today’s docker, each layer of the image will get re-uploaded in its entirety to `docker.io/A/Z` even though all of the content already exists on `docker.io` at `docker.io/OtherUser/testRepo`.

This specification offers an improvement to the performance of pushing and pulling by effectively mounting the shared layers of `docker.io/OtherUser/testRepo` to `docker.io/A/Z` instead of forcing the layers to be re-uploaded. The following section describes the changes that will need to be made in order for this optimization to be possible.

## Tracking Layer References
In order for Cross-Repository Push to work, the daemon must keep track of accessed repositories that reference each layer. This information can be kept in a JSON file called “sources”, which is stored in `/var/lib/docker/graph/<layername>/`

The process of writing layer information to the `sources` file will be implemented in `/docker/graph/graph.go`

### Structure of the `sources` File
The sources file is formatted as a structured list of JSON entries. Each entry contains two fields. The first is the repository in which the layer was referenced, and the second field is a timestamp representing the last time the referenced repository contained the given layer.

For instance, a sources file might look like the following:
`[{“docker.io/myname/myapp”, 8-11-2015-15:23},
 {“docker.io/myfriend/myapp, 8-8-2015-14:35}]`

In the above example, the daemon recorded that the layer was present in the repository `docker.io/myfriend/myapp` at `8-8-2015-14:35`, and also present in the repository `docker.io/myname/myapp` at `8-11-2015-15:23`

Note that each entry in the map is written as a fully qualified repository.

#### Preventing Read-write race conditions
The daemon uses file locking to prevent read-write race conditions.  On read, the goroutine grabs a read lock, which blocks other goroutines from writing to the file, but allows other goroutines to also grab the read lock and read.  On write, the goroutine attempts to take the write lock.  If no other goroutines hold a read lock, then the goroutine grabs the write lock, which blocks all other goroutines from reading or writing.  If other goroutines hold read locks on the file, the goroutine must wait until they all release the locks before taking the write lock and writing to the file.  

#### Concurrency
With cross-repository push, if two clients attempt to pull the same layer (layer A), they will pull in sequence, as the docker daemon currently only allows one client to pull layer A at a time.  If they both attempt to cross-repository push the same layer to different repositories, they succeed.

#### Purpose of the Timestamp Field
In a large enterprise solution, where many users are pushing and pulling against a common base image, it is possible for the sources file to grow out of bounds. Additionally, there is no guarantee that any reference in the sources file is valid. The `timestamp` field is used to address these issues by removing stale entries from the list after a configurable-by-the-user number of days. 

Therefore, the timestamp serves three primary purposes:
  - It bounds the size of the sources file, limiting wasted disk space
  - It keeps the sources file clean and prevents broken references from being sent to the registry
  - It limits the number of reference items that must be sent to the registry in the HTTP request

## API
The API for a cross repository mount will follow the format proposed in the original proposal, [#634](https://github.com/docker/distribution/issues/634)

## Legacy Fallback
Cross Repository Push should not replace the existing method of pulling and pushing from a repository. In the event of a failure or inability to perform a cross repository push, the system should fallback to the traditional upload scheme.

## Additional Notes
This specification refers to cross repository mounting within a single registry. It is not concerned with cross-registry-cross-repository mounting. However, the code should be designed so that cross-registry-cross-repository can be added in the future. With this goal in mind, all references to a repository should be fully qualified.

cc @vbatts @ncdc @shishir-a412ed @jwhonce @rhatdan @alecbenson

**Comment by dmp42** [comment]
+ @docker/distribution-maintainers

**Comment by stevvooe** [comment]
@jramseye This seems mostly to address the daemon implementation. What changes to the distribution interfaces do you plan to make to support this? Will we require any? Can we leverage some of the work from proxy caching? What configuration parameters will there be on the registry side?

I'd like to see this addition include a proper abstraction for managing "pulling" a layer from elsewhere. We may want to leverage the proxy flow, but I'd like to see a plan for the registry modifications.

**Comment by alecbenson** [comment]
@stevvooe 
Can we start with discussing what is here for now? @jramseye an I are still figuring things out on the distribution side and will be updating this as we sort it out. 

That being said, it would be great to get some feedback on what is here for now while the rest is being figured out.

**Comment by RichardScothern** [comment]
Thanks for the proposal @alecbenson .  Can you clarify a few points for me?

How is the sources file built initially?  i.e. in the case of starting a docker daemon with existing pulled repositories.  Is it a blocking or asynchronous operation?

How do you choose a repo from which to mount in the case where the digest is in multiple repos on the same registry?  Do you try others in the case that the blob was soft deleted?  Do you synchronize the source file with the registry state? (remove the entry where the delete happened)

How exactly does the timestamp limit the number of reference items that must be sent to the registry in the HTTP request.  And how does it prevent broken references from being sent to the registry?  Could you give a couple of examples?

**Comment by stevvooe** [comment]
@alecbenson Sure, we can. Here are my big questions:

What does the actual sources data structure look like? Please be very specific and don't be afraid to use a Go struct. Much of the behavior will stem from the quality of the data structure.

Does this behavior sit at the tag store level or the graph store level? These are individual abstractions and we need to be clear where this data belongs.

My questions about the modifications to the distribution interface still stands. Actual, cross-repository push will be implemented in `registry/client` before making it the daemon. We will not accept a patch that implements the cross-repository push hacked into the daemon. Let me know if you need guidance or ideas but I would recommend starting here: http://godoc.org/github.com/docker/distribution. This addition may simply be a new method that transfers a blob from one endpoint to another.

**Comment by alecbenson** [comment]
@stevvooe @RichardScothern Thanks! These are all great questions and I will address as many as I can provide answers to.

>How is the sources file built initially? i.e. in the case of starting a docker daemon with existing pulled repositories. Is it a blocking or asynchronous operation?

When an image is pushed, if no sources file exists, then it should be created. These older images that exist prior to the creation of the sources file will just get pushed up using a legacy push. However, when the images are pushed, then the sources file will be created and an entry will be added for those images and the layers of each. At that point, any subsequent pushes/pulls will be done via a cross repo push, if possible. 

> Is it a blocking or asynchronous operation?

Sorry -- by 'it' are you referring to the creation of the sources file, or just a cross repository push in general?

>How do you choose a repo from which to mount in the case where the digest is in multiple repos on the same registry? Do you try others in the case that the blob was soft deleted? 

I think the best solution here is to simply work down the list until a mount is successful. If all of the sources fail, then it should revert back to a legacy push.

>Do you synchronize the source file with the registry state? (remove the entry where the delete happened)

Yes, in the event that a source file entry is invalid, then it should be removed from the sources file.
 
>How exactly does the timestamp limit the number of reference items that must be sent to the registry in the HTTP request. And how does it prevent broken references from being sent to the registry? Could you give a couple of examples?

It doesn't necessarily *limit* the number of items that get sent to the registry. That was probably bad wording on our part. It does however, keep the `sources` file from becoming bloated with entries that are old and therefore more likely to be out of use.

> What configuration parameters will there be on the registry side?

There should be a flag to disable cross repo pushing on registries that do not want to use the optimization. 

> Does this behavior sit at the tag store level or the graph store level? These are individual abstractions and we need to be clear where this data belongs.

My thoughts are that this belongs more in the graph store, but as I get more familiar with distribution, I'll be able to give a more informed answer to this one. 

>What does the actual sources data structure look like? Please be very specific and don't be afraid to use a Go struct. Much of the behavior will stem from the quality of the data structure.

Yes, the sources data should certainly be unmarshaled into a struct. The struct should contain, at the very least, the repository source and the timestamp. I hope this answer isn't too vague for now, but as @jramseye dive into this, we'll have a clearer picture.

**Comment by RichardScothern** [comment]
Thanks for answering my questions

> Sorry -- by 'it' are you referring to the creation of the sources file, or just a cross repository push in general?

The 'it' was generating the sources file - I was wondering if one would be created before any push operations.  You answered it above.

**Comment by stevvooe** [comment]
> My thoughts are that this belongs more in the graph store, but as I get more familiar with distribution, I'll be able to give a more informed answer to this one.

Agreed. This is likely the right level. We should avoid the tag store for this and really focus on the layer aspect.

> Yes, the sources data should certainly be unmarshaled into a struct. The struct should contain, at the very least, the repository source and the timestamp. I hope this answer isn't too vague for now, but as @jramseye dive into this, we'll have a clearer picture.

That's sufficient. Could you or @jramseye drop a straw man into the description?

**Comment by dmcgowan** [comment]
I agree with the `source` file approach. I think we will need to store the digest in this file as well. Currently we rely on a file called `checksum`, however this value can get overwritten on any given registry push despite our current efforts to keep this stable. But it is also possible that two repositories refer to the same layer id with different digests simply because they were pushed from different hosts. Until we can flip the storage model to be by digest, we should keep this mapping here for each repository.  The `mount` value described in the API is intended to be the digest as referenced by the `from` repository.

**Comment by jramseye** [comment]
@stevvooe Sounds good.  We'll keep it in the graph store, then.  Sorry, what do you mean by "drop a straw man"?

@dmcgowan We chose to omit the digest from the file, as we expect that soon we'll have a stable digest.  If that is not the case, we can switch to storing the digest as well. That's an extra piece of information to keep track of, so we'd rather not store it if unnecessary.  While definitely possible, the odds of two different repos referring to the same layer ID by different digests is low, and seems an acceptable tradeoff to having to store less information. But if the consensus is to store the digest ID as well, we can switch our proposed system.

**Comment by stevvooe** [comment]
@jramseye Show us what the data structure might look like:

```
type Source struct {
    Name string `"json:something,omitempty"`
    Foo int 
}
```

Let's have something concrete to discuss.

"Show me your data structures and I'll show you your code."

**Comment by jramseye** [comment]
@stevvoe Got it, thanks.  
Yes, I think it would be something like that.   @shishir-a412ed, do you have any thoughts on what the structure should be?  I"m thinking something along the lines of:
``` 
type Source struct {
    Repo string 
    Timestamp Time 
}
```
at the very least.

**Comment by stevvooe** [comment]
@jramseye Is "repository" enough to correctly resolve the image location? How would this work in the face of abstraction? For example, how would one correctly set it.

Rather than having a timestamp, would it make sense to have them in LRU order? The entire file has to be rewritten on update, anyways.

Should we keep track of the digest of that the layer was pulled under?

What about something like this:

```go
type Source struct {
    Repository string 
    RegistryURL string
    Digest digest.Digest
    Timestamp time.Time // creation time of the particular record
}
```

Are there any other data structures that we could integrate this with?

**Comment by shishir-a412ed** [comment]
@stevvooe 

I would be taking over @jramseye work on cross repository push .
I am working on a more detailed specification that I ll post by tomorrow. Hopefully that should answer your questions. N we can take the discussion from there.

Shishir

**Comment by stevvooe** [comment]
@shishir-a412ed Great!

**Comment by mattmoor** [comment]
Can you be sure to crosslink the new proposal here when you do, so those of us following this thread can make the jump to the new proposal?

**Comment by ncdc** [comment]
Do you all want this to move to docker/docker and be a PR so everyone can comment on individual lines?

**Comment by vbatts** [comment]
After mulling on this further, the contentious points around locking and freshness of timestamps, I think a better point will be to drive the actual requirement of what those implementation detail were attempting to solve.

* parent child relationship of layers - such that a tagged image and traverse the tree of children layers/images to determine named leaves (images that include that parent image)
* determining freshest, mutually known, names child/image that includes a common parent image - such that this fresh image name can be used when talking to a registry endpoint, for ACL tokens

Is this correct?

**Comment by ncdc** [comment]
What I am concerned about is this scenario:

1. `docker pull centos:7`
1. One of the layers you get has v1 id `c852f6d61e65`
1. Create a new image based on `centos:7` and push it to `repo1:latest`
1. Wait 2 months
1. Tag `repo1:latest` as `anotherrepo:latest` and push it
1. If we evict entries from `sources` whose timestamps are older than e.g. 30 days, we'll end up erasing the entry from layer `c852f6d61e65`'s `sources` file that says the layer came from `centos:7`, resulting in a full push to `anotherrepo` when in fact we could do a cross-repo mount

**Comment by ncdc** [comment]
The purpose of the timestamp was/is to keep the `sources` file from growing without bounds.

There's also the question that @dmcgowan brought up yesterday as to whether there's some way to intelligently pick the entry (or entries) that should be included as request params to the registry, instead of always sending all of them.

**Comment by vbatts** [comment]
but what i'm saying is that if that relationship exists, and the child image has a timestamp itself of "last_push" or similar, then there is no need for a `sources` file per layer.

**Comment by ncdc** [comment]
I see what you're saying. So what would the flow be when I try to cross mount `c852f6d61e65` when pushing to `anotherrepo`?

**Comment by vbatts** [comment]
if there is a proper relationship, then a step up and traverse for freshest parent or sibling would be quick operation. Then you would have a minimal image name on that registry endpoint, that has been pushed and includes that id

**Comment by vbatts** [comment]
@dmp42 @dmcgowan @stevvooe ^^ this is really driving at an ACL/auth handler case.
As an auth handler for cross-repo mount, how can I find the middle ground on freshest sibling of a digest and not have to be sent an unbounded list of image names?

**Comment by mattmoor** [comment]
"growing without bounds"

Do we really feel as though this would grow to an unrealistic size ever?

I'd say keep it all, but this should only trigger on push/pull and not simply tagging events.

**Comment by ncdc** [comment]
On an OpenShift build node, potentially. I think we should walk through some scenarios on our end to see if it's really going to be an issue, or if I'm just worrying prematurely.

**Comment by vbatts** [comment]
@mattmoor the point is that if a common parent, is built upon N leaf images and possibly pushed to 1 or more registries R, then with the common parent `sources` file approach, the list length worse case would be N*R+1 for it's own reference. For high iteration build hosts, that could be large.

Then for each push, that _whole_ set of names for an endpoint need be included, that would be a query parameter of <image names> * N long.

Alternatively, to make this ACL assertion, we only need to send the _latest_ name of freshest sibling, then this long list goes away, and the sent list is nominal.

**Comment by ncdc** [comment]
Assuming the freshest sibling still exists as a valid image repository...

**Comment by shishir-a412ed** [comment]
@ncdc 
In your example above both `repo1:latest` and `anotherrepo:latest` would be doing a cross-repo mount rather than a full legacy push. We are using timestamp to sort entries, but entries will only be deleted after the specified threshold (In my proposal that is 300: this value is open for discussion). 

So if that layer `c852f6d61e65` is associated with 400 different entries. 300 will be kept based on which repos were most recently accessed.

**Comment by ncdc** [comment]
@shishir-a412ed I'm referring to the proposal above, namely

> The timestamp field is used to address these issues by removing stale entries from the list after a configurable-by-the-user number of days.

**Comment by mattmoor** [comment]
Sorry, it took me a second to grok what you were saying.  I get it now, high degree base images have a ton of sources (one for every place a derivative has been pushed).

I guess I am still skeptical that size here would be that much of a problem.

That said, making it a flag with a default of ~300 would keep virtually anyone from hitting it outside of pathological cases like this, so I could live with it.

**Comment by ncdc** [comment]
@mattmoor and actually, we're proposing that the sources file would only have entries added for full pushes, not for cross-repo mounts. So that should help alleviate the size, assuming we can send along the first repo in each registry to receive a full push.

**Comment by vbatts** [comment]
man. I wish github could have threaded comments, not stacked.

1. yes, a "cross-repo push" will be separate from a "legacy push"
2. perhaps a sane watermark of how many image names to push may solve the ACL concern with a buck-shot of the most fresh image names, but not an unbounded list
3. the storage of them in a per layer `sources` files is the sloppy part, that i'd like to see done away with, by a proper tree relationship

**Comment by vbatts** [comment]
(sorry i didn't finish my thought)

3.b. with this fat sources done away with, the need for troubling locking also should go away

**Comment by BrianBland** [comment]
@shishir-a412ed I'd like to breathe new life back into this proposal, and I'm willing to take over the implementation.

@ncdc To clarify what you said, are we only going to update the sources file on full (non-cross-repo) pushes as well as pulls? I'd expect users to pull from the same repositories they do a cross-repo push for, so this would be sort of equivalent to updating the sources on each push/pull operation. Or is it only updated on each full push?

@stevvooe that source struct looks to be sufficient

**Comment by aaronlehmann** [comment]
Closing - cross repository push support on the registry side was merged in #1269.
